### PR TITLE
Bug 4253: ssl_bump prevents from accessing some web contents

### PR DESCRIPTION
**Pull request by chtsanti**
Surrogate-Capability should only be sent on accel traffic, not
on bumped traffic.

This is a Measurement Factory project

### CODE REVIEW

**Comment by chtsanti** [comment]
This is also fixes bug 4880 (it is duplicate of 4253)

**Comment by yadij** [comment]
Can we not move the switchedToHttps() exceptional code to prepareTransparentURL() and use that far simpler function?

**Comment by rousskov** [comment]
I assume Amos is suggesting, for switchedToHttps() requests, to call prepareTransparentURL() instead of calling prepareAcceleratedURL(). I do not know whether both functions would produce identical URLs in SslBump context, but, when answering Amos question, please keep in mind that only prepareAcceleratedURL() sets flags.accel, and there may still be SslBump code that relies on that flag to be set.

**Comment by chtsanti** [comment]
> Can we not move the switchedToHttps() exceptional code to prepareTransparentURL() and use that far simpler function?

Well, these are the questions I wanted to avoid when I proposed an one line change solution!
Now I am a little lost in cases.

The sslBumped requests are not accelarator requests but they are not exactly transparent, so neither the prepareTransparentURL can be used.
Actually the prepareAcceleratedURL builds correct URL for sslBump only if the Host header exist. The prepareTransparentURL will produce correct URL only if Host header exist to.

I conclude to make a new function prepareTlsSwitchingURL. If the port field is missing from URL in some cases is it a huge problem?  Probably can retrieved if required, just may require a little code.

**Comment by rousskov** [comment]
> Amos: Can we not move the switchedToHttps() exceptional code to prepareTransparentURL() and use that far simpler function?

> Christos: these are the questions I wanted to avoid when I proposed an one line change solution!

I am confused by the _combination_ of your reaction to that question and your proposal to add prepareTlsSwitchingURL(): AFAICT, the short answer to Amos' question is "yes" if prepareTransparentURL() does what SslBump needs. The short answer is "no" if prepareTransparentURL() does not do what SslBump needs. And if we cannot really tell what SslBump needs exactly because there are many complex and/or unknown cases, then writing prepareTlsSwitchingURL() to do what SslBump needs (as you propose) sounds like a rather dangerous path!

> I conclude to make a new function prepareTlsSwitchingURL. 

In general, writing prepareTlsSwitchingURL() could be a very good first step. If the result looks very similar to one of the two existing functions mentioned above, then we can avoid code duplication by merging new and old code. If all three functions are very different, then the needs of their callers are probably very different, and all three functions deserve to co-exist (they can still reuse common code, of course). However, if we do not know what prepareTlsSwitchingURL() should do, then this lets-write-and-then-decide logic is not really applicable because we cannot write that function.

BTW, how does the current "Surrogate-Capability should not be sent on bumped traffic" PR description relate to the current PR code changes? Does `flags.accel` control nothing but sending Surrogate-Capability?

**Comment by yadij** [comment]
@rousskov, setting flags.accel enables the logic for handling Surrogate-Control. The bug 4880 is a server brokenly emitting that header without Squid sending the Surrogate-Capability header. Both bugs and several others are separate side effects of the flag being set.

**Comment by yadij** [comment]
@christos, I can go with having a third function for now. But I do expect that you will find it to be almost identical to the existing transparent one.

**Comment by chtsanti** [comment]
> > Christos: these are the questions I wanted to avoid when I proposed an one line change solution!
> 
> I am confused by the _combination_ of your reaction to that question ...

OK, I was tried to avoid looking more on this issue. The two prepare*() functions, even if it is clear what they do, they looked a little complex with the first look!

> 
> BTW, how does the current "Surrogate-Capability should not be sent on bumped traffic" PR description relate to the current PR code changes? Does `flags.accel` control nothing but sending Surrogate-Capability?

My sense is that if the flags.accel flag is not set then squid will not try to process surrogate-control replies. But I did not test.

> @yadij , setting flags.accel enables the logic for handling Surrogate-Control. The bug 4880 is a server brokenly emitting that header without Squid sending the Surrogate-Capability header. Both bugs and several others are separate side effects of the flag being set.

I am not sure that the reported server is broken. In my tests, the site appeared as broken did not enable esi when I am not sending the Surrogate-Capability header.


>  @yadij   I can go with having a third function for now. But I do expect that you will find it to be almost identical to the existing transparent one.

Not exactly. Because if the Host is missing the "prepareTransparentURL" uses the  ConnStateData::clientConnection->local to build url. This is OK for transparent SSL bumping but not when we are bump normal CONNECT requests.

**Comment by chtsanti** [comment]
I added the ConnStateData::prepareTlsSwitchingURL to generate URLs for ssl bumped requests. It is ConnStateData because it uses private members of ConnStateData.
I add the  ConnStateData::tlsConnectPort to hold the port on remote server for bumped requests. It is the only way to has port for all the cases.
I also changed the prepareTransparentURL and prepareAcceleratedURL functions. PLease see if you agree with the changes.

File: src/client_side.cc [hunk 1]
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
...
-        const int url_sz = hp->requestUri().length() + 32 + Config.appendDomainLen +
-                           strlen(host);
-        http->uri = (char *)xcalloc(url_sz, 1);
         const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s" SQUIDSBUFPH,
-                 SQUIDSBUFPRINT(scheme), host, SQUIDSBUFPRINT(hp->requestUri()));
-        debugs(33, 5, "TRANSPARENT HOST REWRITE: " << http->uri);
-    } else {
+        const int url_sz = scheme.length() + strlen(host) + hp->requestUri().length() + 32;
+        uri = (char *)xcalloc(url_sz, 1);
```
Comment (yadij) [inline src/client_side.cc]: casting

File: src/client_side.cc [hunk 2]
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
...
         debugs(33, 5, "ACCEL VPORT REWRITE: *_port IP + vport=" << vport);
         /* Put the local socket IP address as the hostname, with whatever vport we found  */
-        const int url_sz = hp->requestUri().length() + 32 + Config.appendDomainLen;
-        http->uri = (char *)xcalloc(url_sz, 1);
-        http->getConn()->clientConnection->local.toHostStr(ipbuf,MAX_IPSTRLEN);
+        conn->clientConnection->local.toHostStr(ipbuf,MAX_IPSTRLEN);
         const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s:%d" SQUIDSBUFPH,
+        const int url_sz = scheme.length() + sizeof(ipbuf) + url.length() + 32;
+        char *uri = (char *)xcalloc(url_sz, 1);
```
Comment (yadij) [inline src/client_side.cc]: casting

File: src/client_side.cc [hunk 3]
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
...
-        http->uri = (char *)xcalloc(url_sz, 1);
         char vportStr[32];
         vportStr[0] = '\0';
         if (vport > 0) {
             snprintf(vportStr, sizeof(vportStr),":%d",vport);
         }
         const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s%s" SQUIDSBUFPH,
+        const int url_sz = scheme.length() + strlen(conn->port->defaultsite) + sizeof(vportStr) + url.length() + 32;
+        char *uri = (char *)xcalloc(url_sz, 1);
```
Comment (yadij) [inline src/client_side.cc]: casting

File: src/client_side.cc [hunk 4]
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
                 host = thost;
             }
         } // else nothing to alter port-wise.
-        const int url_sz = hp->requestUri().length() + 32 + Config.appendDomainLen + strlen(host);
-        http->uri = (char *)xcalloc(url_sz, 1);
         const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s" SQUIDSBUFPH, SQUIDSBUFPRINT(scheme), host, SQUIDSBUFPRINT(url));
-        debugs(33, 5, "ACCEL VHOST REWRITE: " << http->uri);
+        const int url_sz = scheme.length() + strlen(host) + url.length() + 32;
+        char *uri = (char *)xcalloc(url_sz, 1);
```
Comment (yadij) [inline src/client_side.cc]: casting

File: src/client_side.cc [hunk 5]
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
...
+        Must(sslConnectHostOrIp.size());
+        SBuf useHost;
+        if (!tlsClientSni().isEmpty())
+            useHost = tlsClientSni();
+        else
+            useHost.assign(sslConnectHostOrIp.rawBuf(), sslConnectHostOrIp.size());
+
+        const SBuf &scheme = AnyP::UriScheme(transferProtocol.protocol).image();
+        const int url_sz = scheme.length() + useHost.length() + hp->requestUri().length() + 32;
+        uri = (char *)xcalloc(url_sz, 1);
```
Comment (yadij) [inline src/client_side.cc]: casting

**Comment by rousskov** [comment]
@eduard-bagdasaryan, please check why Anubis is not updating this (or the next `M-cleared-for-merge`) PR. I am probably missing something, but it feels like Anubis is stuck.

**Comment by eduard-bagdasaryan** [comment]
> it feels like Anubis is stuck

It got stuck on PR30. The reason is unclear, I suspect it hung during an API call of getting statuses from GitHub. So I have to I restarted it, and now PR306 is being processed.

**Comment by rousskov** [comment]
Glad I asked! Thank you for addressing the immediate problem. Long-term, does Anubis need API call timeouts? I thought we already had those...

**Comment by eduard-bagdasaryan** [comment]
> It got stuck on PR30.

I found the true reason of this problem and fixed.

File: src/client_side.cc [hunk 6]
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
...
+
+static char *
+prepareTransparentURL(ConnStateData * conn, const Http1::RequestParserPointer &hp)
+{
+    // TODO Must() on URI !empty when the parser supports throw. For now avoid assert().
+    if (!hp->requestUri().isEmpty() && hp->requestUri()[0] != '/')
+        return nullptr; /* already in good shape */
+
+    char *uri = buildUrlFromHost(conn, hp);
+    if (!uri){
```
[the remark this answers is not in the document]
Comment (chtsanti) [reply src/client_side.cc]: ok

File: src/client_side.cc [hunk 7]
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
...
-        const int url_sz = hp->requestUri().length() + 32 + Config.appendDomainLen;
-        http->uri = (char *)xcalloc(url_sz, 1);
         static char ipbuf[MAX_IPSTRLEN];
-        http->getConn()->clientConnection->local.toHostStr(ipbuf,MAX_IPSTRLEN);
-        const SBuf &scheme = AnyP::UriScheme(http->getConn()->transferProtocol.protocol).image();
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s:%d" SQUIDSBUFPH,
+        conn->clientConnection->local.toHostStr(ipbuf,MAX_IPSTRLEN);
+        const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
+        const int url_sz = sizeof(ipbuf) + hp->requestUri().length() + 32;
+        uri = (char *)xcalloc(url_sz, 1);
```
[the remark this answers is not in the document]
Comment (chtsanti) [reply src/client_side.cc]: ok

File: src/client_side.cc [hunk 8]
```diff
@@ -1142,18 +1140,16 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
...
     char *host = NULL;
-    if (tryHostHeader && (host = hp->getHeaderField("Host"))) {
+    if (vhost && (host = hp->getHeaderField("Host"))) {
         debugs(33, 5, "ACCEL VHOST REWRITE: vhost=" << host << " + vport=" << vport);
         char thost[256];
         if (vport > 0) {
             thost[0] = '\0';
             char *t = NULL;
-            if (host[strlen(host)] != ']' && (t = strrchr(host,':')) != NULL) {
+            if (host[strlen(host) - 1] != ']' && (t = strrchr(host,':')) != NULL) {
```
[the remark this answers is not in the document]
Comment (chtsanti) [reply src/client_side.cc]: ok

File: src/client_side.h [hunk 9]
```diff
@@ -397,6 +398,7 @@ class ConnStateData : public Server, public HttpControlMsgSink, private Independ
 
     /// The SSL server host name appears in CONNECT request or the server ip address for the intercepted requests
     String sslConnectHostOrIp; ///< The SSL server host name as passed in the CONNECT request
+    unsigned short tlsConnectPort; ///< The TLS server port name as passed in the CONNECT request
```
Comment (yadij) [inline src/client_side.h]: number
```suggestion
    unsigned short tlsConnectPort; ///< The TLS server port number as passed in the CONNECT request
```

File: src/client_side.cc [hunk 10]
```diff
@@ -3051,6 +3100,7 @@ ConnStateData::switchToHttps(HttpRequest *request, Ssl::BumpMode bumpServerMode)
     assert(!switchedToHttps_);
 
     sslConnectHostOrIp = request->url.host();
+    tlsConnectPort =  request->url.port();
```
Comment (yadij) [inline src/client_side.cc]: whitespace
```suggestion
    tlsConnectPort = request->url.port();
```

File: src/client_side.cc [hunk 11]
```diff
@@ -1342,9 +1387,11 @@ parseHttpRequest(ConnStateData *csd, const Http1::RequestParserPointer &hp)
      *  - remote interception with PROXY protocol
      *  - remote reverse-proxy with PROXY protocol
      */
-    if (csd->transparent()) {
+    if (csd->switchedToHttps()) {
+        http->uri = csd->prepareTlsSwitchingURL(hp);
+    } else if (csd->transparent() || csd->switchedToHttps()) {
```
Comment (yadij) [inline src/client_side.cc]: This switchedToHttps() condition can never be true, if it was the if-statement above would have run instead.
```suggestion
    } else if (csd->transparent()) {
```

File: src/client_side.h (hunk 9 above)
Comment (chtsanti) [reply src/client_side.h]: addressed at 19f00737

File: src/client_side.cc (hunk 10 above)
Comment (chtsanti) [reply src/client_side.cc]: addressed at 19f00737

File: src/client_side.cc (hunk 11 above)
Comment (chtsanti) [reply src/client_side.cc]: addressed at 19f00737

**Comment by chtsanti** [comment]
@yadij 
I have addressed all change requests in your review. Please re-review.

**Comment by yadij** [comment]
Updating to master to trigger the Jenkins re-build. Earlier failures were node errors, not code related.

**Pull request closed without merging**